filemanager/miniomanager.go

```diff
@@ -124,54 +125,58 @@ GetObjectNameFromLocation gets the object name/key name from the object location
 	https://minio-endpoint/bucket-name/key1 - >> key1
 	http://minio-endpoint/bucket-name/key2 - >> key2
 */
-func (manager *minioManager) GetObjectNameFromLocation(location string) (string, error) {
+func (m *minioManager) GetObjectNameFromLocation(location string) (string, error) {
 	var baseURL string
-	if manager.config.UseSSL {
+	if m.config.UseSSL {
 		baseURL += "https://"
 	} else {
 		baseURL += "http://"
 	}
-	baseURL += manager.config.EndPoint + "/"
-	baseURL += manager.config.Bucket + "/"
+	baseURL += m.config.EndPoint + "/"
+	baseURL += m.config.Bucket + "/"
 	return location[len(baseURL):], nil
 }
 
-func (manager *minioManager) GetDownloadKeyFromFileLocation(location string) string {
+func (m *minioManager) GetDownloadKeyFromFileLocation(location string) string {
 	parsedUrl, err := url.Parse(location)
 	if err != nil {
 		fmt.Println("error while parsing location url: ", err)
 	}
 	trimedUrl := strings.TrimLeft(parsedUrl.Path, "/")
-	return strings.TrimPrefix(trimedUrl, fmt.Sprintf(`%s/`, manager.config.Bucket))
+	return strings.TrimPrefix(trimedUrl, fmt.Sprintf(`%s/`, m.config.Bucket))
 }
 
-func (manager *minioManager) objectUrl(objectName string) string {
+func (m *minioManager) objectUrl(objectName string) string {
 	protocol := "http"
-	if manager.config.UseSSL {
+	if m.config.UseSSL {
 		protocol = "https"
 	}
-	return protocol + "://" + manager.config.EndPoint + "/" + manager.config.Bucket + "/" + objectName
+	return protocol + "://" + m.config.EndPoint + "/" + m.config.Bucket + "/" + objectName
 }
 
-func (manager *minioManager) getClient() (*minio.Client, error) {
-	var err error
-	if manager.client == nil {
-		manager.client, err = minio.New(manager.config.EndPoint, &minio.Options{
-			Creds:  credentials.NewStaticV4(manager.config.AccessKeyID, manager.config.SecretAccessKey, ""),
-			Secure: manager.config.UseSSL,
-		})
-		if err != nil {
-			return &minio.Client{}, err
+func (m *minioManager) getClient() (*minio.Client, error) {
+	m.clientOnce.Do(func() {
+		if m.client == nil {
+			m.client, m.clientErr = minio.New(m.config.EndPoint, &minio.Options{
+				Creds:  credentials.NewStaticV4(m.config.AccessKeyID, m.config.SecretAccessKey, ""),
+				Secure: m.config.UseSSL,
+			})
+			if m.clientErr != nil {
```

Comment:
We don't need to initialize to an empty struct if there is an error since we do the error check when we return?

Reply:
I don't know the exact reason why this was done so I kept the original logic. What I'm trying to do here is to just fix the data race. We can do a follow up later and look into rudder server and check all places where this is used to see if we can return `nil` in case of `error`. Sounds good?

Reply:
Sure